tuxemon/states/control/__init__.py

```diff
@@ -0,0 +1,129 @@
+from tuxemon.constants import paths
+from tuxemon.locale import T
+from tuxemon.menu.interface import MenuItem
+from tuxemon.menu.menu import PopUpMenu
+from tuxemon.platform.events import PlayerInput
+from tuxemon.state import State
+
+ControlStateObj = Callable[[], object]
+cfg_file = config.TuxemonConfig(paths.USER_CONFIG_PATH).cfg
+config = prepare.CONFIG
+
+def reload_control_config():
+    config.a = cfg_file.get("controls", "a")
+    config.b = cfg_file.get("controls", "b")
+    config.up = cfg_file.get("controls", "up")
+    config.left = cfg_file.get("controls", "left")
+    config.down = cfg_file.get("controls", "down")
+    config.right = cfg_file.get("controls", "right")
+    config.back = cfg_file.get("controls", "back")
+
+class SetKeyState(PopUpMenu):
+    """This state is responsible for setting the input keys"""
+    shrink_to_items = True
+
+    def startup(self, **kwargs: Any) -> None:
+        self.input = kwargs["input"]
+        super().startup(**kwargs)
+
+        menu_items_map = (
+            T.translate("options_new_input_key0"),
+            T.translate("options_new_input_key1")
+        )
+
+        for key in menu_items_map:
+            label = T.translate(key).upper()
+            image = self.shadow_text(label)
+            item = MenuItem(image, label, None, None)
+            self.add(item)
+
+    def process_event(self, event: PlayerInput) -> Optional[PlayerInput]:
+        pressed_key = None
+        if pygame.key.get_pressed()[pygame.K_UP]: pressed_key = "up"
+        if pygame.key.get_pressed()[pygame.K_DOWN]: pressed_key = "down"
+        if pygame.key.get_pressed()[pygame.K_RIGHT]: pressed_key = "right"
+        if pygame.key.get_pressed()[pygame.K_LEFT]: pressed_key = "left"
+        for k in range(len(pygame.key.get_pressed())):
+            if pygame.key.get_pressed()[k]: pressed_key = k
+        if pressed_key is not None and event.pressed or event.value == "": 
+            # TODO: fix or rewrite PlayerInput
+            # event.value is being compared here since sometimes the value just returns an empty 
+            # string and event.pressed doesn't return True when a key is being pressed
+            if isinstance(pressed_key, str): pressed_key_str = pressed_key
+            if isinstance(pressed_key, int): pressed_key_str = pygame.key.name(pressed_key)
+
+            cfg_file.set("controls", self.input, pressed_key_str)
+            with open(paths.USER_CONFIG_PATH, "w") as fp:
+                cfg_file.write(fp)
+
+            self.client.replace_state(ControlState)
+            self.close()
+
+        super().process_event(event)
+
+class ControlState(PopUpMenu[ControlStateObj]):
+    """This state is responsible for the option menu"""
+
+    escape_key_exits = True
+    shrink_to_items = True
+
+    def startup(self, **kwargs: Any) -> None:
+        # TODO: update the menu once a control key has been changed
+        # TODO: remove the need to restart the game for changes to take place
```

Comment:
To apply this without restarting, you need to update the existing input map for the keyboard.
The game client has an input manager, which aggregates the various input methods and produces virtual input events.  So you will need to access the game client, the input manager, find the keyboard input, finally, modify the input map.  The game client can be accessed using the "local session", i've included an example of it below

https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/client.py#L93
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/platform/events.py#L93
https://github.com/Tuxemon/Tuxemon/blob/development/tuxemon/platform/platform_pygame/events.py#L45

Reply:
I tried your solutions, but they don't seem to work unless I'm missing something. Any other ideas on how to fix it? I've sent a new commit.

Reply:
My bad, I missed something.